HPCC-34406 Create phonenumber plugin

Michael-Gardner opened the pull request:
<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

Commit:
HPCC-34406 Create phonenumber plugin

Signed-off-by: Michael Gardner <michael.gardner@lexisnexisrisk.com>

Michael-Gardner commented:
@mckellyln @ghalliday I have a README.md and a testing/regress/ecl/phonenumber.ecl  test file that I want to include as well, but I'm not certain that I should commit them because they include random phone numbers.  The phone numbers aren't tied to any person with any of our data. They're just stuff I made up but might be real in some cases, in order to get the library to accept that they're valid numbers.

In the readme I can swap everything over to a 1 555 012 3456 US format that isn't valid, just for the purposes of  documentation.  Not sure what to do about the test file though.

Michael-Gardner commented on plugins/phonenumber/phonenumber.cpp:
```diff
@@ -0,0 +1,185 @@
+
+IPluginContext * parentCtx = NULL;
+
+ECL_PHONENUMBER_API bool ECL_PHONENUMBER_CALL checkValidity(ICodeContext *_ctx, size32_t lenNumber, const char *number, size32_t lenCountryCode, const char *countryCode)
+{
+    const i18n::phonenumbers::PhoneNumberUtil* phoneUtil = i18n::phonenumbers::PhoneNumberUtil::GetInstance();
+    i18n::phonenumbers::PhoneNumber phoneNumber;
+    try
+    {
+        std::string numberStr(number, lenNumber);
+        std::string countryCodeStr(countryCode, lenCountryCode);
+        phoneUtil->Parse(numberStr, countryCodeStr, &phoneNumber);
+        return phoneUtil->IsValidNumber(phoneNumber);
```
Not a horrible idea, working on that now.

GordonSmith commented on plugins/phonenumber/lib_phonenumber.ecllib:
```diff
@@ -0,0 +1,27 @@
+############################################################################## */
+
+// Service definition
+EXPORT phonenumber := SERVICE : plugin('phonenumber'), library('phonenumber'), namespace('HPCCPhoneNumber'), CPP, PURE
```
By convention it looks like we use `plugin('phonenumberplugin')` and `namespace('phonenumber')`

GordonSmith commented on plugins/phonenumber/phonenumber.hpp:
```diff
@@ -0,0 +1,48 @@
+    ECL_PHONENUMBER_API void ECL_PHONENUMBER_CALL phonenumberType(ICodeContext * _ctx, size32_t &lenResult, char *&result, size32_t lenNumber, const char *number, size32_t lenCountryCode, const char *countryCode);
+    ECL_PHONENUMBER_API void ECL_PHONENUMBER_CALL regionCode(ICodeContext * _ctx, size32_t &lenResult, char *&result, size32_t lenNumber, const char *number, size32_t lenCountryCode, const char *countryCode);
+    ECL_PHONENUMBER_API unsigned ECL_PHONENUMBER_CALL countryCode(ICodeContext * _ctx, size32_t lenNumber, const char *number, size32_t lenCountryCode, const char *countryCode);
+}
```
No need to declare functions here

Michael-Gardner replied:
Removed

GordonSmith commented on plugins/phonenumber/lib_phonenumber.ecllib:
```diff
@@ -0,0 +1,27 @@
+    BOOLEAN isValidNumber(CONST STRING phonenumber, CONST STRING countryCode) : cpp,action,context,fold,entrypoint='checkValidity';
+    STRING getType(CONST STRING phonenumber, CONST STRING countryCode) : cpp,action,context,fold,entrypoint='phonenumberType';
+    STRING getRegionCode(CONST STRING phonenumber, CONST STRING countryCode) : cpp,action,context,fold,entrypoint='regionCode';
+    INTEGER getCountryCode(CONST STRING phonenumber, CONST STRING countryCode) : cpp,action,context,fold,entrypoint='countryCode';
```
@ghalliday is there any prefernce between `CONST STRING` and `CONST VARSTRING`?

GordonSmith commented on plugins/phonenumber/phonenumber.cpp:
```diff
@@ -0,0 +1,184 @@
+{
+    const i18n::phonenumbers::PhoneNumberUtil* phoneUtil = i18n::phonenumbers::PhoneNumberUtil::GetInstance();
+    std::string numberStr(number);
+    std::string countryCodeStr(countryCode);
```
Should include the length in the constructor

GordonSmith reviewed (changes requested):
General comment:  could the catch (...) be replaced with something more explicit?

Also can you add some regression tests, see:
* testing/regress/ecl/h3.ecl
* testing/regress/ecl/key/h3.xml

Michael-Gardner commented on plugins/phonenumber/phonenumber.cpp:
```diff
@@ -0,0 +1,184 @@
+        default:
+            resultStr = "INVALID";
+            break;
+        }
```
@GordonSmith I was able to add an Enum into the code so that phonenumberType returns the enum.  I can't seem to get it to work properly in the ECL though.  I have a work around that I tested. The enum is explicitly declared as an integer. The expected return for getType (entrypoint=phonenumberType) is now an Integer.  Attempts at making it take the ecl phonenumber_type ENUM(integer) just wasn't working. Regardless of how I tried to wrap the calls and do the return.  Gavin said I should leave it as is with the Integer and we can update it to an Enum at a later date.

GordonSmith commented on plugins/phonenumber/CMakeLists.txt:
```diff
@@ -0,0 +1,83 @@
+        )
+
+        target_link_libraries(phonenumber
+            PUBLIC
```
Should be PRIVATE

GordonSmith commented on plugins/phonenumber/CMakeLists.txt:
```diff
@@ -0,0 +1,83 @@
+                libphonenumber::phonenumber
+                absl::base
+                absl::strings
+                absl::log
```
Is absl needed here?

GordonSmith commented on plugins/phonenumber/lib_phonenumber.ecllib:
```diff
@@ -0,0 +1,27 @@
+    limitations under the License.
+############################################################################## */
+
+EXPORT phonenumber_type := ENUM(INTEGER, FIXED_LINE=0, MOBILE=1, FIXED_LINE_OR_MOBILE=2, TOLL_FREE=3, PREMIUM_RATE=4, SHARED_COST=5, VOIP=6, PERSONAL_NUMBER=7, PAGER=8, UAN=9, VOICEMAIL=10, UNKNOWN=11);
```
nitpick: change to
```ecl
EXPORT phonenumber_type := ENUM(INTEGER,
    FIXED_LINE=0, 
    MOBILE, 
    FIXED_LINE_OR_MOBILE, 
    TOLL_FREE, 
    PREMIUM_RATE, 
    SHARED_COST, 
    VOIP, 
    PERSONAL_NUMBER, 
    PAGER, 
    UAN, 
    VOICEMAIL, 
    UNKNOWN
);
```

GordonSmith commented on plugins/phonenumber/phonenumber.cpp:
```diff
@@ -0,0 +1,156 @@
+    const i18n::phonenumbers::PhoneNumberUtil* phoneUtil = i18n::phonenumbers::PhoneNumberUtil::GetInstance();
+    std::string numberStr(number, lenNumber);
+    std::string countryCodeStr(countryCode, lenCountryCode);
+    phoneUtil->Parse(numberStr, countryCodeStr, &phoneNumber);
```
Should return the ErrorType

GordonSmith commented on plugins/phonenumber/phonenumber.cpp:
```diff
@@ -0,0 +1,156 @@
+    catch (...)
+    {
+        return false;
+    }
```
Is try catch needed? should return ErrorType?

GordonSmith commented on plugins/phonenumber/lib_phonenumber.ecllib:
```diff
@@ -0,0 +1,27 @@
+
+EXPORT phonenumber_type := ENUM(INTEGER, FIXED_LINE=0, MOBILE=1, FIXED_LINE_OR_MOBILE=2, TOLL_FREE=3, PREMIUM_RATE=4, SHARED_COST=5, VOIP=6, PERSONAL_NUMBER=7, PAGER=8, UAN=9, VOICEMAIL=10, UNKNOWN=11);
+
+
```
Add ErrorType enum

GordonSmith commented on plugins/phonenumber/phonenumber.cpp:
```diff
@@ -0,0 +1,156 @@
+
+    try
+    {
+        parseNumber(lenNumber, number, lenCountryCode, countryCode, phoneNumber);
```
Check for error

GordonSmith commented on plugins/phonenumber/phonenumber.cpp:
```diff
@@ -0,0 +1,156 @@
+    catch (...)
+    {
+        regionCodeStr = "";
+    }
```
Check for error and is try/catch needed?

GordonSmith commented on plugins/phonenumber/phonenumber.cpp:
```diff
@@ -0,0 +1,156 @@
+    catch (...)
+    {
+        return 0;
+    }
```
Check for error and no need for try/catch.

GordonSmith commented on plugins/phonenumber/lib_phonenumber.ecllib:
```diff
@@ -0,0 +1,27 @@
+    BOOLEAN isValidNumber(CONST STRING phonenumber, CONST STRING countryCode) : cpp,action,context,fold,entrypoint='checkValidity';
+    INTEGER getType(CONST STRING phonenumber, CONST STRING countryCode) : cpp,action,context,fold,entrypoint='phonenumberType';
+    STRING getRegionCode(CONST STRING phonenumber, CONST STRING countryCode) : cpp,action,context,fold,entrypoint='regionCode';
+    INTEGER getCountryCode(CONST STRING phonenumber, CONST STRING countryCode) : cpp,action,context,fold,entrypoint='countryCode';
```
Rather than 4 API calls, it would be better to have one, which return a RECORD of the parsed items.  Otherwise the phone number will get parsed 4 times to extract the full info.

GordonSmith commented on testing/regress/ecl/phonenumber.ecl:
```diff
@@ -0,0 +1,64 @@
+
+TestRecord := RECORD
+  STRING phonenumber;
+	STRING countryCode;
```
indentation

GordonSmith commented on testing/regress/ecl/phonenumber.ecl:
```diff
@@ -0,0 +1,64 @@
+  STRING phonenumber;
+	STRING countryCode;
+  BOOLEAN valid;
+	STRING description;
```
indentation

GordonSmith commented on testing/regress/ecl/phonenumber.ecl:
```diff
@@ -0,0 +1,64 @@
+                                               SELF.pRegionCode := lib_phonenumber.phonenumber.getRegionCode(LEFT.phonenumber, LEFT.countryCode),
+                                               SELF := LEFT
+                                               )
+                        );
```
indentation

GordonSmith commented on vcpkg.json.in:
```diff
@@ -111,6 +111,10 @@
+            "name": "libphonenumber",
+            "platform": "@VCPKG_PHONENUMBER@"
+        },
+	    {
```
indentation

Michael-Gardner commented:
@Gordon I moved everything into a parseNumber() function.  Setup a record for phonenumber_data.  The function parseNumber() now serializes the data and stores it on the heap using MemoryBuffer.  I optimized for size when we're dealing with the country codes and the phonenumber_type enum.  Also updated the Readme and phonenumber.ecl test file.

GordonSmith commented on testing/regress/ecl/phonenumber.ecl:
```diff
@@ -0,0 +1,63 @@
+                                                SELF.pType := lib_phonenumber.phonenumber.parseNumber(LEFT.phonenumber, LEFT.countryCode)[1].lineType,
+                                                SELF.pCountryCode := lib_phonenumber.phonenumber.parseNumber(LEFT.phonenumber, LEFT.countryCode)[1].countryCode,
+                                                SELF.pRegionCode := lib_phonenumber.phonenumber.parseNumber(LEFT.phonenumber, LEFT.countryCode)[1].regionCode,
+                                                SELF := LEFT));
```
Not sure the ECL above is correct?

Michael-Gardner commented on plugins/phonenumber/lib_phonenumber.ecllib:
```diff
@@ -0,0 +1,52 @@
+    BOOLEAN valid;
+    phonenumber_type lineType;
+    STRING regionCode{MAXLENGTH(5)};
+    INTEGER2 countryCode;
```
Added

Michael-Gardner commented on plugins/phonenumber/phonenumber.cpp:
```diff
@@ -0,0 +1,129 @@
+    const i18n::phonenumbers::PhoneNumberUtil* phoneUtil = i18n::phonenumbers::PhoneNumberUtil::GetInstance();
+    std::string numberStr(number, lenNumber);
+    std::string countryCodeStr(countryCode, lenCountryCode);
+    throwOnError(phoneUtil->Parse(numberStr, countryCodeStr, &phoneNumber));
```
Done

GordonSmith commented on testing/regress/ecl/phonenumber.ecl:
```diff
@@ -0,0 +1,91 @@
+        lib_phonenumber.phonenumber_type.VOIP => 'VoIP Service',
+        lib_phonenumber.phonenumber_type.PREMIUM_RATE => 'Premium Rate',
+        'Other/Unknown'
+    );
```
Add this helper logic to the `ecllib` file as its useful!

Michael-Gardner commented on plugins/phonenumber/phonenumber.cpp:
```diff
@@ -0,0 +1,120 @@
+    std::string formattedNumber;
+    phoneUtil->Format(phoneNumber, i18n::phonenumbers::PhoneNumberUtil::PhoneNumberFormat::E164, &formattedNumber);
+    
+    bool isValid = phoneUtil->IsValidNumber(phoneNumber);
```
No. So a number that is not rout-able can still pass the parse step.  isValid means that it's a routable number for that region.  Example would be a US number with 555 prefix.  It will parse w/ NO_PARSING_ERROR, return a region and country code and an unknown line type. But it will return false for being valid (rout-able).

Michael-Gardner commented on plugins/phonenumber/phonenumber.cpp:
```diff
@@ -0,0 +1,120 @@
+    std::string numberStr(number, lenNumber);
+    std::string countryCodeStr(countryCode, lenCountryCode);
+
+    errorType err = static_cast<errorType>(phoneUtil->Parse(numberStr, countryCodeStr, &phoneNumber));
```
From reviewing the code, it won't cause a crash. It looks like the PhoneNumber object will be set to sane defaults.  A good example in the output is with this French number that fails parsing.  Here's the output.

<img width="1497" height="177" alt="image" src="https://github.com/user-attachments/assets/b96111cb-7b2e-46d2-a18e-1bcc83ee1ca0" />

The error is `enum i18n::phonenumbers::PhoneNumberUtil::ErrorType::TOO_LONG_NSN = 5`

The country code is set to 0, region is set to ZZ (default unknown), linetype is 11 which is unknown.

So it won't crash the code, but we do save processing time if I check the return code and just pass back the default PhoneNumber object immediately instead of calling the other functions on it.

I'll make a change.

GordonSmith commented on plugins/phonenumber/lib_phonenumber.ecllib:
```diff
@@ -0,0 +1,62 @@
+    NOT_A_NUMBER,
+    TOO_SHORT_AFTER_IDD, // Too short after International Direct Dialing prefix
+    TOO_SHORT_NSN, // Too short to be a phone number
+    TOO_LONG_NSN // Too long to be a phone number
```
What does NSN stand for?

GordonSmith commented on plugins/phonenumber/phonenumber.cpp:
```diff
@@ -0,0 +1,98 @@
+    const i18n::phonenumbers::PhoneNumberUtil* phoneUtil = i18n::phonenumbers::PhoneNumberUtil::GetInstance();
+    std::string numberStr(number, lenNumber);
+    std::string countryCodeStr(countryCode, lenCountryCode);
+    // Set default values that would normally returend from calls to the PhoneNumber object if the number is invalid
```
Typo

GordonSmith commented on testing/regress/ecl/phonenumber.ecl:
```diff
@@ -0,0 +1,193 @@
+  {'03 1234 5678', 'JP', true, 'Japanese with leading zero'},
+  {'010 1234 5678', 'CN', true, 'Chinese with leading zero'},
+  {'01 42 68 53 00', 'FR', true, 'French with leading zero'},
+  {'06 698 80000', 'IT', true, 'Italian with leading zero'}
```
I would add some tests where the two letter country code is omitted?

Michael-Gardner replied:
Good idea.

GordonSmith reviewed (changes requested):
Looks good - just a typo and some minor comments.

Michael-Gardner commented:
@GordonSmith fixed, squashed

GordonSmith reviewed (commented):
Looks good, you will need to add the missing /opt/HPCCSystems/testing/regress/ecl/key/phonenumber.xml file for the regression tests to pass

Michael-Gardner commented:
@GordonSmith fixed the phonenumber.xml file so it passes the regression with the expected outputs.  Also added a line to the testing/regression/README.rst file that would have saved me a headache.

<img width="831" height="82" alt="image" src="https://github.com/user-attachments/assets/140a7c7b-49ec-4731-92a0-fd382a42e375" />

GordonSmith commented on testing/regress/ecl/key/phonenumber.xml:
```diff
@@ -0,0 +1,328 @@
+ <Row><phonenumber>+49 30 123</phonenumber><country_code></country_code><tr_valid>false</tr_valid><description>Invalid German international without country context</description><number>+4930123</number><error>0</error><valid>true</valid><linetype>0</linetype><regioncode>DE</regioncode><countrycode>49</countrycode></Row>
+</Dataset>
+<Dataset name='invalidNumbers'>
+ <Row><phonenumber>+1 555 012 3456</phonenumber><country_code>US</country_code><tr_valid>false</tr_valid><description>US with international prefix</description><number>+15550123456</number><error>0</error><valid>false</valid><linetype>11</linetype><regioncode>ZZ</regioncode><countrycode>1</countrycode></Row>
```
Why is this result invalid?  There are a few of these which have error==0 but are invalid?

GordonSmith reviewed (commented):
Question about some of the invalid tests

ghalliday commented on plugins/phonenumber/CMakeLists.txt:
```diff
@@ -0,0 +1,75 @@
+
+project(phonenumber)
+
+if(PHONENUMBER)
```
Is this ever true?  Other plugins have a USE_XX test.

GordonSmith replied:
This is correct.  The general pattern for plugins is `INCLUDE_XXX` or `SUPPRESS_XXX`

Michael-Gardner replied:
So the USE_XX test is generally if there's a third party library that we sometimes couldn't support.

If you look at h3, kafka, or mysqlembed, they all use the IF(PLUGINNAME) style.  If you look at Memcached, it uses the IF(PLUGINNAME AND USE_XX) style. I believe that was because we wanted to turn off memcached on certain build targets because it wasn't supported.  Now that we have vcpkg and bundle dependent libraries, things are simpler and we don't really need to have the USE_XX style.

ghalliday reviewed (commented):
When I tried to build this locally it did not succeed.  Any idea what I am doing wrong?

ghalliday reviewed (approved):
@Michael-Gardner Looks really good.  One question/sugestion.

Michael-Gardner commented:
@ghalliday Thanks for the suggestion, I'll go ahead and test that.

I noticed a failure when running the phonenumber.ecl tests on thor. Going to figure out why that's happening and push a fix along with your suggested changes.

ghalliday reviewed (approved):
Once @GordonSmith approves please squash
